runtime/v1/shim/service.go

```diff
@@ -503,47 +503,42 @@ func (s *Service) processExits() {
 	}
 }
 
-func (s *Service) allProcesses() []process.Process {
+func (s *Service) checkProcesses(e runc.Exit) {
+	var p process.Process
 	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	res := make([]process.Process, 0, len(s.processes))
-	for _, p := range s.processes {
-		res = append(res, p)
+	for _, proc := range s.processes {
+		if proc.Pid() == e.Pid {
+			p = proc
+			break
+		}
 	}
-	return res
-}
-
-func (s *Service) checkProcesses(e runc.Exit) {
-	for _, p := range s.allProcesses() {
-		if p.Pid() != e.Pid {
-			continue
+	s.mu.Unlock()
+	if p == nil {
+		log.G(s.context).Infof("process with id:%d wasn't found", e.Pid)
+		return
+	}
+	if ip, ok := p.(*process.Init); ok {
+		shouldKillAll, err := shouldKillAllOnExit(s.bundle)
```

Comment:
this part will be changed by https://github.com/containerd/containerd/pull/4057. just hold it right now.